docs/specs/code/TRANSACTION_MEMO.md

```diff
@@ -0,0 +1,239 @@
+
+Currently, the user-flow for a standard Yoroi wallet already uses a **spending password** to decrypt the encrypted master key of the wallet. On hardware wallets, however, the spending password is not required as the wallet handles the private keys internally. That being said, we propose to use the existing spending password to derive a new **memo password** for symmetric encryption/decryption. We prefer not to use the *spending password* itself to encrypt/decrypt memos because we want to avoid compromising the security of the spending password in any way. 
+
+We define the memo password as the result of `Blake2b` applied to the concatenation of the wallet account public key and the spending password. The memo password is encrypted using the spending password and saved to local storage.
+
+In terms of user interaction in standard Yoroi wallets, the user will continue to use the spending password to send transactions with memos without noticing any change, as the same password will decrypt the master key for create the transaction and decrypt the memo password used to encrypt memos. To read the memo of an existing transaction, the user will have to introduce his/her spending password.
+
+For hardware wallets, we propose to imitate the user-flow used in standard wallets. That is, the user will have to introduce his/her spending password at the time of sending a transaction **if and only if** he/she decides to add a memo. Similarly, the user will be asked for the spending password when reading a memo associated with a transaction.
+
+Regarding the signing process, notice that `ChaCha20Poly1305` includes a MAC which can be used to verify *integrity* and *authenticity* of the data. For the sake of simplicity, we consider that signing the memo to prove *ownership* is not necessary, as having a valid memo password implies having access to the spending password and the specific wallet (with its account public key).
+
+A Proof-of-Concept can be found [here](https://repl.it/repls/DimgreyImmaterialDebuggers).
+
+We use the Cardano encrypted message format in the same way as before. Similarly, the ciphertext is the result of applying 'password_ecnrypt' to the content of the memo.
+
+**Cardano encrypted message**
+```
+16(
+  [
+    / protected / h'' / {
+      \ is_ascii \ is_ascii:false \
+    } / ,
+    / unprotected / {
+      \ enc_type \ enc_type:1 \
+      \ version \ version:1 \
+    } /, 
+    / ciphertext / h'56294e41264b6eeaf9952f7e452ef05bdbeb8b3985a5ca05a88e848ce989483edb3783d92ba0c04cbdd4911825a3c0aebab032513ca4caea553d741434582969d356cd380587361b'
+  ]
+)
+```
+
+The above output is saved to a file named after the `Blake2b` hash of the corresponding transaction id and uploaded to the external storage. To decrypt we do the opposite process.
+
+### Public key vs User password
+
+We make a brief comparison of both methods advantages/disadvantages.
+
+| # | Description           | Public key | User password |
```

Comment:
as Shin said, we can't use `user password` since the password could change

Reply:
That's right. Just for the record, we can't use a **spending-password** based strategy as the user could use different spending passwords across distinct instances of the same wallet.